File: 17. Palindrome Partitioning (Min Cuts).py
# 17. Palindrome Partitioning (Min Cuts)
# Topic: Palindrome Partitioning
# Description: Minimize cuts needed to partition a string into palindromic substrings using memoization.
# Input: single line string s
# Output: minimum cuts

import sys
from functools import lru_cache

def is_pal(s: str, i: int, j: int) -> bool:
    while i < j:
        if s[i] != s[j]:
            return False
        i += 1; j -= 1
    return True

if __name__ == '__main__':
    line = sys.stdin.read().splitlines()
    if not line:
        sys.exit(0)
    s = line[0].rstrip('\n')
    n = len(s)

    @lru_cache(maxsize=None)
    def min_parts(i: int) -> int:
        if i >= n:
            return 0
        best = 10**9
        for j in range(i, n):
            if is_pal(s, i, j):
                best = min(best, 1 + min_parts(j+1))
        return best

    parts = min_parts(0)
    # number of cuts = partitions - 1
    print(max(0, parts - 1))

# Memoization process started at min_parts decorated with lru_cache.
# Time Complexity: O(n^2) with palindrome precomputation; current naive check O(n^3)
# Space Complexity: O(n)

File: 17. Palindrome Partitioning (Min Cuts).cpp
// 17. Palindrome Partitioning (Min Cuts)
// Topic: Palindrome Partitioning
// Description: Given string s, compute minimum cuts to partition into palindromic substrings using
// memoization. Input: s (one line) Output: minimum cuts

#include <bits/stdc++.h>
using namespace std;

bool isPal(const string& s, int i, int j)
{
    while (i < j)
    {
        if (s[i++] != s[j--])
            return false;
    }
    return true;
}

int minCuts(const string& s, int i, vector<int>& memo)
{
    int n = s.size();
    if (i >= n)
        return 0;
    if (memo[i] != -1)
        return memo[i];
    int best = INT_MAX / 2;
    for (int j = i; j < n; ++j)
    {
        if (isPal(s, i, j))
        {
            int cost = 1 + minCuts(s, j + 1, memo);
            if (cost < best)
                best = cost;
        }
    }
    memo[i] = best;
    return memo[i];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s;
    if (!getline(cin, s))
        return 0;
    int n = s.size();
    vector<int> memo(n, -1);
    int ans = minCuts(s, 0, memo);
    if (ans == INT_MAX / 2)
        cout << 0 << '\n';
    else
        cout << max(0, ans - 1) << '\n';
}

// Memoization process started in minCuts where memo[i] caches minimal partitions from i; palindrome
// checks used to prune. Time Complexity: O(n^2) average with precomputation; current naive
// palindrome check gives O(n^3) worst-case Space Complexity: O(n)
